### PR TITLE
[Enhancement] Support prune table function arg columns

### PR DESCRIPTION
**Pull request by kangkaisen**
## Problem Summary:
for the following sql, we only need to san column v3, needn't scan v4, v5

```
        String sql = "select count(t1.v7) from t0_array, " +
                "unnest(t0_array.v3, t0_array.v4, t0_array.v5) as t1(v7, v8, v9);";
```

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneTableFunctionColumnRule.java [hunk 1]
```diff
@@ -40,21 +42,61 @@ public PruneTableFunctionColumnRule() {
     public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
         LogicalTableFunctionOperator logicalTableFunctionOperator = (LogicalTableFunctionOperator) input.getOp();
         ColumnRefSet requiredOutputColumns = context.getTaskContext().getRequiredColumns();
-
         List<ColumnRefOperator> newOuterCols = Lists.newArrayList();
+
         for (ColumnRefOperator col : logicalTableFunctionOperator.getOuterColRefs()) {
             if (requiredOutputColumns.contains(col.getId())) {
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneTableFunctionColumnRule.java:48]: this may prune all cloumns.

Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneTableFunctionColumnRule.java:48]: I have considered this point. unnest function supports empty columns.

```
class Unnest final : public TableFunction {
public:
    std::pair<Columns, UInt32Column::Ptr> process(TableFunctionState* state) const override {
        if (state->get_columns().empty()) {
            return {};
        }
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTableFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneTableFunctionColumnRule.java|35|35|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 52 / 52 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneEmptyTableFunctionRule.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTableFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneTableFunctionColumnRule.java|33|33|100.00%|[]|

File: fe/fe-core/src/test/resources/sql/optimized-plan/lateral.sql [hunk 2]
```diff
@@ -8,8 +8,7 @@ TABLE FUNCTION (unnest)
 [sql]
 select v1 from tarray, lateral unnest(v3)
```
Comment (packy92) [inline fe/fe-core/src/test/resources/sql/optimized-plan/lateral.sql:9]: It seems these two plans are not equal. The old plan will expand rows in tarray, like v1 = 3, v3= [1,2,3] will expand to v1= 3, unnest = 1; v1= 3, unnest = 2; v1= 3, unnest = 3;

**Comment by packy92** [comment]
`select count(t1.v7) from t0_array, unnest(t0_array.v3, t0_array.v4, t0_array.v5) as t1(v7, v8, v9);`

**Pull request closed without merging**